nixos test ref/05 class ify test driver

blaggacao opened the pull request:
nixos/test/test-driver: Class-ify the test driver

This commit encapsulates the involved domain into classes and
defines explicit and typed arguments where untyped dicts where used.

It preserves backwards compatibility through legacy wrappers.

/cc @nlewo next in series

Commit:
nixos/test/test-driver: Class-ify the test driver

This commit encapsulates the involved domain into classes and
defines explicit and typed arguments where untyped dicts where used.

It preserves backwards compatibility through legacy wrappers.

blaggacao commented:
```
@GrahamcOfBorg test bittorrent
@GrahamcOfBorg test login
@GrahamcOfBorg test boot.biosCdrom
@GrahamcOfBorg test boot.biosNetboot
@GrahamcOfBorg test boot.biosUsb
@GrahamcOfBorg test boot.uefiCdrom
@GrahamcOfBorg test boot.uefiNetboot
@GrahamcOfBorg test boot.uefiUsb
@GrahamcOfBorg test installer.bcache
@GrahamcOfBorg test installer.btrfsSimple
@GrahamcOfBorg test installer.btrfsSubvolDefault
@GrahamcOfBorg test installer.btrfsSubvols
@GrahamcOfBorg test installer.encryptedFSWithKeyfile
@GrahamcOfBorg test installer.grub1
@GrahamcOfBorg test installer.luksroot
@GrahamcOfBorg test installer.lvm
@GrahamcOfBorg test installer.separateBoot
@GrahamcOfBorg test installer.simple
@GrahamcOfBorg test installer.swraid
@GrahamcOfBorg test installer.fsroot
```

blaggacao commented:
@Ma27 Can you please check if this re-implementation for `nixos-build-vms` satisfies the spec? I cloudn't cleanly incorporate your recent changes into the new 'control flow' in the `main` function.

blaggacao commented:
```
rg create_machine
nixos/tests/hibernate.nix
71:          return create_machine(

nixos/tests/installer.nix
99:      def create_machine_named(name):
100:          return create_machine({**default_flags, "name": name})
138:      machine = create_machine_named("boot-after-install")
196:      machine = create_machine_named("rebuild-switch")
216:      machine = create_machine_named("boot-after-rebuild-switch")
226:      machine = create_machine_named("clone-default-config")
246:      machine = create_machine_named("clone-alternate-config")

nixos/tests/common/ec2.nix
60:        machine = create_machine({"startCommand": start_command})

nixos/tests/boot.nix
36:            machine = create_machine(${machineConfig})
77:            machine = create_machine(${machineConfig})
```

```
@GrahamcOfBorg test common.ec2
@GrahamcOfBorg test hibernate
```

Ma27 commented:
Thanks for the heads-up, will take a look later today.

Ma27 committed:
nixos/build-vms: fix eval

Ma27 committed:
nixos/test-driver: fix graphics for VM

`-nographic` should only be set if no display is available. Otherwise,
tools such as `nixos-build-vms(8)` will never open up a graphical
display.

Ma27 commented:
@blaggacao I took the liberty to push two small fixes for issues I encountered (`nixos-build-vms` didn't evaluate, no QEMU display was shown when interactively starting a VM), I hope that's OK :) 

The functionality seems OK now (if you want to test it on your own in the future, you can also do `NIX_PATH=nixpkgs=$(pwd) nixos-build-vms vm.nix && ./result/bin/nixos-run-vms` in your nixpkgs checkout, `nixos-build-vms` will automatically select the correct `build-vms.nix` then).

I haven't taken a closer look at the rest of the changes, I'd prefer to leave that to e.g. @tfc since they IIRC wrote most of this initially. If the VM tests and the review passes, this should be good to go IMHO.

blaggacao commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -245,110 +229,252 @@ def _perform_ocr_on_screenshot(
+            f"{self._cmd}"
+            f" -monitor unix:{monitor_socket_path}"
+            f" -chardev socket,id=shell,path={shell_socket_path}"
+            f"{display_opts}"
```
Note-to-self: Check where `quemu_ops` went here during rebase?!?

nlewo replied:
Nowhere! but they have been restored when i run all NixOS tests.

blaggacao commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -245,110 +229,252 @@ def _perform_ocr_on_screenshot(
+            {
+                "TMPDIR": str(state_dir),
+                "SHARED_DIR": str(shared_dir),
+                "USE_TMPDIR": "1",
```
@tfc I couldn't really make sense of the neccessity of this env variable. We might take a critical look at it.

blaggacao commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -245,110 +229,252 @@ def _perform_ocr_on_screenshot(
+    that also satisfies the requirement of the BaseStartCommand.
+    These nix command have the particular charactersitic that the
+    machine name can be extracted out of them via a regex match.
+    (Admittedly a _very_ implicit contract, evtl. TODO fix)
```
@roberth I'd like to get your eyes back on deck here, since this is our angle of approach (for not to say "attack" :wink:) to solve the additional restriction on hostnames over RFC dns lables.

We can't fix that without fixing the way that the hostnames are recovered from within the test driver. (What about FQDN's with clashing host labels?)

Once fixed, this entire class should go away.

roberth replied:
If I remember correctly; on the NixOS side, all we need is [system.name](https://search.nixos.org/options?channel=unstable&show=system.name&from=0&size=50&sort=relevance&query=system.name) to be our python identifier and `networking.hostName` to do its usual thing.

However, we need to get both names. I don't think the interface between python and Nix is set up for that, judging by the awkward parsing. Do we generate a nice json on the Nix side yet?

nlewo replied:
We could change the `startScripts` argument type from `[Path]` to `[(MachineName, Path)]` where `MachineName` is a `String`. The real machine name could then be injected at buildtime by Nix to the Python driver. The Python driver could then sanitize this machine name to create a valid Python variable name (`. -> _`, `pass -> pass_`, ... ). (This would also allow us to create a nice helping message to indicate the machine name transformation that has been applied.) 
Moreover, the real machine name could still be used as keys in the `machines` dict.

I could address this in a followup PR.

roberth replied:
Let's avoid full-on "sanitization" though. It'd be something that the caller has to learn and something we may have to break in the future. If we require the caller to use python identifiers in the `machines` attrset, they're still free to set `network.hostName`.
We can default the `network.hostName` to the python identifier, with `_` swapped for `-`. If you consider that sanitization, then touché, but it is well delineated because python identifiers are quite restrictive. Hostnames, in contrast, can have both `.` and `-`, making the reverse translation non-injective. That would make me sad :(. At least we can take this simplified view, because we're only talking about a default value. Restrictions on the first character make the picture more complicated, which is why it's great to defer the final naming decisions to the user, who already has to deal with restrictions anyway. And those are the restrictions that are inherent to _their_ test, not the restrictions we try to predict.

nlewo replied:
My bad, i actually forgot the machine has a name in the node attribute set! So, i agree we don't need to sanitize it since the caller could have to choose a name which is also a valid Python identifier.

However, instead of parsing the outpath, i think we could still pass the machine name as script arguments.

blaggacao replied:
I think, we should probably starting on resolving this issue immediately after getting this PR through since it promises another cleanup of the code path.

Maybe indeed the best here is to use the machine name of the `nodes` attribute as exported symbols, but also hold fqdn in a `machines` dict.

roberth replied:
:+1: on scoping this out.

It'd be nice to have a new NixOS option `system.labels = mkOption {type = json;}` for the "sole" purpose of being written to a file `$toplevel/labels.json`. A nice interface for passing data from `config` to NixOS 'consumers'. The name borrows from k8s, fwiw. Not `system.`_`nixos`_`.labels` because `nix-darwin` should be able to do the same.

blaggacao commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -245,110 +229,252 @@ def _perform_ocr_on_screenshot(
+class LegacyStartCommand(StartCommand):
+    """used in some places to create an ad-hoc machine instead of
+    using nix test instrumentation + module system for that purpose.
+    Legacy.
```
I presume it shall be always preferable to declare a machine configuration through nix functions rather than rely on a opaque driver interface, here.

We might need to discuss a plan to transition away from it's usage via `create_machine`?

blaggacao commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -908,47 +1011,223 @@ def unblock(self) -> None:
         """Make the machine reachable."""
         self.send_monitor_command("set_link virtio-net-pci.1 on")
 
+    def release(self) -> None:
```
@kraftnix would this have any impact on the slow tear downs that you had discovered and reported in the devos chat?

blaggacao commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -908,47 +1011,223 @@ def unblock(self) -> None:
-    with log.nested("waiting for all VMs to finish"):
-        for machine in machines:
-            machine.wait_for_shutdown()
+        # TODO: don't side-effect environment here
+        os.environ[f"QEMU_VDE_SOCKET_{self.nr}"] = str(self.socket_dir)
```
@tfc do you have an idea if/how we can avoid to pipe that over the global environment / state?

tfc replied:
with the current design of the vm scripts, unfortunately not. but our restauration PRs have a history of being very overloaded and so i would really like to scope that out.

blaggacao commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -908,47 +1011,223 @@ def unblock(self) -> None:
+        self.fd = os.fdopen(pty_master, "w")
+        self.fd.write("version\n")
+
+        # TODO: perl version checks if this can be read from
+        # an if not, dies. we could hang here forever. Fix it.
+        assert self.process.stdout is not None
+        self.process.stdout.readline()
```
@tfc Maybe now is the time to fix this?

nlewo replied:
Because the current PR already improves the current implementation, i would suggest to postpone this ;)

blaggacao commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -908,47 +1011,223 @@ def unblock(self) -> None:
+        self.process.terminate()
+
+
+class Driver:
+    """A handle to the driver that sets up the environment
+    and runs the tests"""
+
+    tests: str
+    vlans: List[VLan]
+    machines: List[Machine]
```
@tfc I hope you enjoy the new terseness of the driver class as much as I do :wink:

tfc replied:
i surely do!

nlewo commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -242,113 +229,252 @@ def _perform_ocr_on_screenshot(
-                    + ",werror=report,if=none "
-                    + "-device scsi-hd,drive=hda "
+class LegacyStartCommand(StartCommand):
+    """used in some places to create an ad-hoc machine instead of
```
```suggestion
    """Used in some places to create an ad-hoc machine instead of
```

nlewo commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -242,113 +229,252 @@ def _perform_ocr_on_screenshot(
+class NixStartScript(StartCommand):
+    """A start script from nixos/modules/virtualiation/qemu-vm.nix
+    that also satisfies the requirement of the BaseStartCommand.
+    These nix command have the particular charactersitic that the
```
```suggestion
    These Nix commands have the particular charactersitic that the
```

nlewo reviewed (commented):
I still need a bit of time to review it but for now it looks pretty good!

nlewo commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -908,41 +1011,225 @@ def unblock(self) -> None:
+            run_tests=self.run_tests,
+            join_all=self.join_all,
+            retry=retry,
+            serial_stdout_off=serial_stdout_off,
```
```suggestion
            serial_stdout_off=self.serial_stdout_off,
```

nlewo commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -908,41 +1011,225 @@ def unblock(self) -> None:
+            join_all=self.join_all,
+            retry=retry,
+            serial_stdout_off=serial_stdout_off,
+            serial_stdout_on=serial_stdout_on,
```
```suggestion
            serial_stdout_on=self.serial_stdout_on,
```

nlewo reviewed (commented):
lgtm

blaggacao commented:
@tfc @nlewo Since this time, we actually don't modify the test execution env, while more changed lines of code, this might actually be the less riskier change.

The most "uncelean" codpath remains test that make use of `create_machine`. But those have been instructed for testing here: https://github.com/NixOS/nixpkgs/pull/126713#issuecomment-902446538

I'll rerun the tests after a rebase the rebase I just did.

```
@GrahamcOfBorg test bittorrent
@GrahamcOfBorg test login
@GrahamcOfBorg test boot.biosCdrom
@GrahamcOfBorg test boot.biosNetboot
@GrahamcOfBorg test boot.biosUsb
@GrahamcOfBorg test boot.uefiCdrom
@GrahamcOfBorg test boot.uefiNetboot
@GrahamcOfBorg test boot.uefiUsb
@GrahamcOfBorg test installer.bcache
@GrahamcOfBorg test installer.btrfsSimple
@GrahamcOfBorg test installer.btrfsSubvolDefault
@GrahamcOfBorg test installer.btrfsSubvols
@GrahamcOfBorg test installer.encryptedFSWithKeyfile
@GrahamcOfBorg test installer.grub1
@GrahamcOfBorg test installer.luksroot
@GrahamcOfBorg test installer.lvm
@GrahamcOfBorg test installer.separateBoot
@GrahamcOfBorg test installer.simple
@GrahamcOfBorg test installer.swraid
@GrahamcOfBorg test installer.fsroot
@GrahamcOfBorg test common.ec2
@GrahamcOfBorg test hibernate
```

nlewo commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -1055,44 +1297,20 @@ def test_symbols() -> Dict[str, Any]:
+        tic = time.time()
+        driver.start_all()
+        driver.run_tests()
+        driver.join_all()
```
Why do you `start_all` and `join_all`?
(this is a new behavior and some tests never terminating)

tfc replied:
please don't do this. let's keep the semantics, there is no need to change them in general and also there is no need to have semantic changes mixed with this MR.

nlewo replied:
I removed these calls in a fixup commit.

nlewo committed:
nixos.tests.udisks2: state_dir is now of type pathlib.Path

nlewo committed:
nixos.tests.usbguard: state_dir is now of type pathlib.Path

nlewo commented:
In order to ensure this PR won't break too much tests, i choose a recent commit already built by Hydra to extract the list of successfully built tests. I then rebased this branch on top of this commit and fixed this PR until all tests work.

The chosen Hydra evaluation is https://hydra.nixos.org/eval/1702797 (commit 09cd65b33c5653d7d2954fef4b9f0e718c899743). I then extracted `613` tests (only successful x86_64 tests).

And after several iterations, all of these `613` tests have been successfully built on top of this PR :tada:

@blaggacao I pushed two `fixup` commits containing several fixes that have to be integrated into your original commit (two other commits are fixing tests because of some type changes). Could you rebase your branch (of course, do not hesitate to ask me any question)?

I'm now pretty confident this branch could be merged without breaking too many tests:)
We would still need to keep an eye on Hydra once merged...

nlewo commented:
@Ma27 Did you test the `buildvms` related scripts on top of this PR? And if true, do you consider it's working as expected?

Ma27 commented:
LGTM :+1:

tfc commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -908,41 +1016,225 @@ def unblock(self) -> None:
-    with log.nested("waiting for all VMs to finish"):
-        for machine in machines:
-            machine.wait_for_shutdown()
+    def start(self) -> None:
```
hm... we had that discussion before. i still see no use case for a non-initialized vlan. now after a *lot* of other changes, this comes back. why?

nlewo replied:
Your commit has been cherrypicked into this branch.

Commit:
Re-RAII-ify the NixOS integration test driver's VLAN class.

We have no usecase for manually/selectively starting or stopping VLANs
in integration tests.
By starting and stopping the VLANs with the constructor and destructor
of VLAN objects, we remove the obligation and complexity to maintain
network lifetime separately.

mkenigs commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -242,113 +229,257 @@ def _perform_ocr_on_screenshot(
+class StartCommand:
+    """The Base Start Command knows how to append the necesary
+    runtime qemu options as determined by a particular test driver
+    run. Any such start command is expected to hapily receive and
```
```suggestion
    run. Any such start command is expected to happily receive and
```

mkenigs commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -242,113 +229,257 @@ def _perform_ocr_on_screenshot(
+        self.state_dir = self.tmp_dir / f"vm-state-{self.name}"
+        self.monitor_path = self.state_dir / "monitor"
+        self.shell_path = self.state_dir / "shell"
+        if (not self.keep_vm_state) and self.state_dir.exists():
+            shutil.rmtree(self.state_dir)
+            rootlog.info(f"    -> delete state @ {self.state_dir}")
```
```suggestion
        if not self.keep_vm_state:
            self.cleanup_statedir()
```

nlewo replied:
Fixed.

mkenigs commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -827,13 +934,14 @@ def process_serial_output() -> None:
     def cleanup_statedir(self) -> None:
         if os.path.isdir(self.state_dir):
             shutil.rmtree(self.state_dir)
```
rmtree checks if the path is a directory, and if the path is a file, we might as well throw an exception now rather than waiting for something else to fail (eg `mkdir(path)` in `__init__`)
```suggestion
        shutil.rmtree(self.state_dir)
```

nlewo commented:
@tfc I'm running NixOS tests (during this night) and will approve the PR once they are executed. Otherwise, the PR looks good to me.

nlewo commented:
Locally, only 3 tests are failing and they were also failing on the parent commit of this PR. So, i consider we are safe.

dasJ commented:
Is it possible that the test runner does not print exceptions for subtests anymore since this is merged?
```
Test "timers" failed with error:
clean up
kill machine (pid 9)
(0.00 seconds)
```

blaggacao commented:
@dasJ I would assume that the codepath that is touched is the nested logger, is that correct?

blaggacao commented:
@dasJ I'm sorry you don't approve the service level I _could_ humanely offer at this point im time. Thanks for the fix. You where indeed closer to the issue than me.
